{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "\n",
    "from scipy.stats import ttest_ind\n",
    "\n",
    "# Classifiers\n",
    "from sklearn.ensemble import AdaBoostClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.discriminant_analysis import LinearDiscriminantAnalysis\n",
    "from sklearn.tree import DecisionTreeClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.gaussian_process import GaussianProcessClassifier\n",
    "from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis\n",
    "from xgboost import XGBClassifier\n",
    "from sklearn.naive_bayes import GaussianNB\n",
    "\n",
    "\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.metrics import log_loss, ConfusionMatrixDisplay, average_precision_score, accuracy_score\n",
    "\n",
    "\n",
    "# suppress futurewarnings\n",
    "def warn(*args, **kwargs):\n",
    "    pass\n",
    "import warnings\n",
    "warnings.warn = warn"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing raw data\n",
    "Processing and preparation of training data from the raw data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# READING DATA\n",
    "exprsn_data = pd.read_csv(\"C:/Users/TEJINDER SINGH/Downloads/New folder (13)/esets_renal_exprs.csv\")\n",
    "output_labels = pd.read_csv(\"C:/Users/TEJINDER SINGH/Downloads/kidney_label.csv\")\n",
    "\n",
    "# DATA TRANSPOSE\n",
    "out_t = output_labels.set_index(\"Source Name\").transpose()\n",
    "kidney_data = exprsn_data.set_index(\"Unnamed: 0\").transpose()\n",
    "\n",
    "# SELECTING THE COLUMNS or SAMPLE NAMES\n",
    "data_c = exprsn_data.columns\n",
    "out_c = out_t.columns\n",
    "\n",
    "# GENERATING LABELS FOR DATASET\n",
    "output = pd.DataFrame()\n",
    "for i in data_c:\n",
    "    for j in out_c:\n",
    "        if i == j:\n",
    "            output[i] = out_t[j]\n",
    "output = output.T\n",
    "labels = output.pop(\"Characteristics[disease]\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get gene names\n",
    "Saving data and get gene names using annotation package from bioconductor using R programming language"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "kidney_data[\"labels\"] = labels\n",
    "kidney_data.to_csv(r\"C:\\Users\\TEJINDER SINGH\\Downloads\\New folder (13)\\Project Datasets\\kidney_data.csv\", index = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature engineering\n",
    "Selecting and merging data from probes belonging to the same gene by computing their mean."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_common_genes(data):\n",
    "    \"\"\"\n",
    "    Selecting and merging data from probes belonging to the same gene by computing their mean.\n",
    "    \n",
    "    Args:\n",
    "    Dataset containg gene symbols as column names\n",
    "    \n",
    "    Returns:\n",
    "    Dataset containg gene symbols as column names after merging columns with same names/genes\n",
    "    \"\"\"\n",
    "    nan_list = [item for item, frame in data[\"Symbol\"].iteritems() if pd.isnull(frame)]        \n",
    "\n",
    "    for i in nan_list:                                  \n",
    "        data.loc[i,'Symbol'] = data.loc[i,'Unnamed..0']#Adding probe names to the \"Symbols\" column with missing values\n",
    "    data = data.set_index(\"Symbol\").transpose()\n",
    "    data.drop(\"Unnamed..0\", inplace = True)\n",
    "    data = pd.DataFrame(data, dtype = float)\n",
    "    data = data.groupby(data.columns, axis=1).mean()\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Symbol</th>\n",
       "      <th>1007_s_at</th>\n",
       "      <th>1294_at</th>\n",
       "      <th>1552283_s_at</th>\n",
       "      <th>1552388_at</th>\n",
       "      <th>1552401_a_at</th>\n",
       "      <th>1552411_at</th>\n",
       "      <th>1552412_a_at</th>\n",
       "      <th>1552449_a_at</th>\n",
       "      <th>1552563_a_at</th>\n",
       "      <th>1552607_at</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM904985_0014N_Exp.CEL</th>\n",
       "      <td>10.759910</td>\n",
       "      <td>8.168967</td>\n",
       "      <td>6.148286</td>\n",
       "      <td>6.148655</td>\n",
       "      <td>3.526734</td>\n",
       "      <td>6.000679</td>\n",
       "      <td>3.567700</td>\n",
       "      <td>3.420196</td>\n",
       "      <td>4.406512</td>\n",
       "      <td>4.929491</td>\n",
       "      <td>...</td>\n",
       "      <td>6.492485</td>\n",
       "      <td>6.470051</td>\n",
       "      <td>6.311163</td>\n",
       "      <td>5.141446</td>\n",
       "      <td>6.531090</td>\n",
       "      <td>5.263501</td>\n",
       "      <td>8.519608</td>\n",
       "      <td>7.424008</td>\n",
       "      <td>6.131245</td>\n",
       "      <td>7.506218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM904986_0014T_Exp.CEL</th>\n",
       "      <td>10.627069</td>\n",
       "      <td>8.110592</td>\n",
       "      <td>6.080880</td>\n",
       "      <td>5.371208</td>\n",
       "      <td>3.637927</td>\n",
       "      <td>6.378871</td>\n",
       "      <td>3.765375</td>\n",
       "      <td>3.907632</td>\n",
       "      <td>3.826696</td>\n",
       "      <td>4.338781</td>\n",
       "      <td>...</td>\n",
       "      <td>6.286828</td>\n",
       "      <td>7.491875</td>\n",
       "      <td>6.998897</td>\n",
       "      <td>5.153848</td>\n",
       "      <td>6.105484</td>\n",
       "      <td>5.527082</td>\n",
       "      <td>8.349207</td>\n",
       "      <td>8.321408</td>\n",
       "      <td>6.346957</td>\n",
       "      <td>7.314657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM904988_0026N_Exp.CEL</th>\n",
       "      <td>9.912308</td>\n",
       "      <td>7.682694</td>\n",
       "      <td>6.348944</td>\n",
       "      <td>6.496141</td>\n",
       "      <td>3.910959</td>\n",
       "      <td>6.273258</td>\n",
       "      <td>3.618695</td>\n",
       "      <td>3.352324</td>\n",
       "      <td>4.026041</td>\n",
       "      <td>4.973174</td>\n",
       "      <td>...</td>\n",
       "      <td>5.748054</td>\n",
       "      <td>6.875448</td>\n",
       "      <td>6.024658</td>\n",
       "      <td>5.197408</td>\n",
       "      <td>6.384318</td>\n",
       "      <td>5.878829</td>\n",
       "      <td>9.062821</td>\n",
       "      <td>7.490204</td>\n",
       "      <td>6.189373</td>\n",
       "      <td>7.569571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM904989_0026T_Exp.CEL</th>\n",
       "      <td>9.928639</td>\n",
       "      <td>7.883371</td>\n",
       "      <td>5.569622</td>\n",
       "      <td>5.740533</td>\n",
       "      <td>4.167279</td>\n",
       "      <td>6.086869</td>\n",
       "      <td>3.926209</td>\n",
       "      <td>4.000820</td>\n",
       "      <td>3.821508</td>\n",
       "      <td>4.645940</td>\n",
       "      <td>...</td>\n",
       "      <td>6.031481</td>\n",
       "      <td>7.578293</td>\n",
       "      <td>6.753043</td>\n",
       "      <td>5.233582</td>\n",
       "      <td>6.783810</td>\n",
       "      <td>5.051750</td>\n",
       "      <td>8.167005</td>\n",
       "      <td>7.381080</td>\n",
       "      <td>6.078681</td>\n",
       "      <td>7.563371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM904990_0042N_Exp.CEL</th>\n",
       "      <td>9.774650</td>\n",
       "      <td>7.406771</td>\n",
       "      <td>5.963449</td>\n",
       "      <td>6.341914</td>\n",
       "      <td>4.242397</td>\n",
       "      <td>6.336297</td>\n",
       "      <td>4.029943</td>\n",
       "      <td>4.483494</td>\n",
       "      <td>4.035602</td>\n",
       "      <td>4.798371</td>\n",
       "      <td>...</td>\n",
       "      <td>5.964617</td>\n",
       "      <td>6.630055</td>\n",
       "      <td>5.452919</td>\n",
       "      <td>4.918226</td>\n",
       "      <td>5.866362</td>\n",
       "      <td>6.279175</td>\n",
       "      <td>8.704949</td>\n",
       "      <td>7.301923</td>\n",
       "      <td>6.059809</td>\n",
       "      <td>7.520830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM305112.CEL</th>\n",
       "      <td>9.285833</td>\n",
       "      <td>7.404473</td>\n",
       "      <td>6.081948</td>\n",
       "      <td>6.165685</td>\n",
       "      <td>3.623140</td>\n",
       "      <td>7.124676</td>\n",
       "      <td>3.767614</td>\n",
       "      <td>4.018204</td>\n",
       "      <td>4.466480</td>\n",
       "      <td>4.825818</td>\n",
       "      <td>...</td>\n",
       "      <td>6.311605</td>\n",
       "      <td>8.180971</td>\n",
       "      <td>6.332484</td>\n",
       "      <td>5.163339</td>\n",
       "      <td>6.301880</td>\n",
       "      <td>5.545485</td>\n",
       "      <td>7.943559</td>\n",
       "      <td>7.335683</td>\n",
       "      <td>6.315128</td>\n",
       "      <td>7.418185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM305113.CEL</th>\n",
       "      <td>10.256438</td>\n",
       "      <td>8.009185</td>\n",
       "      <td>4.694489</td>\n",
       "      <td>6.140019</td>\n",
       "      <td>3.725265</td>\n",
       "      <td>6.931847</td>\n",
       "      <td>4.055042</td>\n",
       "      <td>4.138074</td>\n",
       "      <td>4.350811</td>\n",
       "      <td>4.730193</td>\n",
       "      <td>...</td>\n",
       "      <td>6.414612</td>\n",
       "      <td>7.152420</td>\n",
       "      <td>5.885697</td>\n",
       "      <td>5.171652</td>\n",
       "      <td>6.137954</td>\n",
       "      <td>5.507664</td>\n",
       "      <td>8.010172</td>\n",
       "      <td>7.625686</td>\n",
       "      <td>5.855927</td>\n",
       "      <td>6.761258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM305114.CEL</th>\n",
       "      <td>11.092388</td>\n",
       "      <td>7.577346</td>\n",
       "      <td>5.547016</td>\n",
       "      <td>5.930863</td>\n",
       "      <td>3.799916</td>\n",
       "      <td>6.136671</td>\n",
       "      <td>3.736061</td>\n",
       "      <td>3.608018</td>\n",
       "      <td>3.942664</td>\n",
       "      <td>4.627653</td>\n",
       "      <td>...</td>\n",
       "      <td>6.173863</td>\n",
       "      <td>8.952424</td>\n",
       "      <td>6.582268</td>\n",
       "      <td>5.146336</td>\n",
       "      <td>6.318175</td>\n",
       "      <td>5.391035</td>\n",
       "      <td>8.517406</td>\n",
       "      <td>8.858774</td>\n",
       "      <td>5.960262</td>\n",
       "      <td>7.712653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM305115.CEL</th>\n",
       "      <td>10.170089</td>\n",
       "      <td>7.499742</td>\n",
       "      <td>5.888226</td>\n",
       "      <td>5.880348</td>\n",
       "      <td>4.193258</td>\n",
       "      <td>6.850084</td>\n",
       "      <td>4.166192</td>\n",
       "      <td>4.287220</td>\n",
       "      <td>3.860280</td>\n",
       "      <td>4.833088</td>\n",
       "      <td>...</td>\n",
       "      <td>6.805837</td>\n",
       "      <td>8.734431</td>\n",
       "      <td>6.018990</td>\n",
       "      <td>5.350754</td>\n",
       "      <td>6.120509</td>\n",
       "      <td>5.969984</td>\n",
       "      <td>8.396805</td>\n",
       "      <td>8.194778</td>\n",
       "      <td>6.022666</td>\n",
       "      <td>7.640047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM305116.CEL</th>\n",
       "      <td>10.051092</td>\n",
       "      <td>8.134782</td>\n",
       "      <td>5.523868</td>\n",
       "      <td>5.861997</td>\n",
       "      <td>3.898903</td>\n",
       "      <td>6.643763</td>\n",
       "      <td>3.875280</td>\n",
       "      <td>3.616558</td>\n",
       "      <td>3.925251</td>\n",
       "      <td>4.884328</td>\n",
       "      <td>...</td>\n",
       "      <td>7.429347</td>\n",
       "      <td>7.343067</td>\n",
       "      <td>6.964285</td>\n",
       "      <td>5.533286</td>\n",
       "      <td>6.539190</td>\n",
       "      <td>5.024315</td>\n",
       "      <td>8.618043</td>\n",
       "      <td>7.417723</td>\n",
       "      <td>5.797618</td>\n",
       "      <td>7.454214</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>347 rows Ã— 32927 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Symbol                   1007_s_at   1294_at  1552283_s_at  1552388_at  \\\n",
       "GSM904985_0014N_Exp.CEL  10.759910  8.168967      6.148286    6.148655   \n",
       "GSM904986_0014T_Exp.CEL  10.627069  8.110592      6.080880    5.371208   \n",
       "GSM904988_0026N_Exp.CEL   9.912308  7.682694      6.348944    6.496141   \n",
       "GSM904989_0026T_Exp.CEL   9.928639  7.883371      5.569622    5.740533   \n",
       "GSM904990_0042N_Exp.CEL   9.774650  7.406771      5.963449    6.341914   \n",
       "...                            ...       ...           ...         ...   \n",
       "GSM305112.CEL             9.285833  7.404473      6.081948    6.165685   \n",
       "GSM305113.CEL            10.256438  8.009185      4.694489    6.140019   \n",
       "GSM305114.CEL            11.092388  7.577346      5.547016    5.930863   \n",
       "GSM305115.CEL            10.170089  7.499742      5.888226    5.880348   \n",
       "GSM305116.CEL            10.051092  8.134782      5.523868    5.861997   \n",
       "\n",
       "Symbol                   1552401_a_at  1552411_at  1552412_a_at  1552449_a_at  \\\n",
       "GSM904985_0014N_Exp.CEL      3.526734    6.000679      3.567700      3.420196   \n",
       "GSM904986_0014T_Exp.CEL      3.637927    6.378871      3.765375      3.907632   \n",
       "GSM904988_0026N_Exp.CEL      3.910959    6.273258      3.618695      3.352324   \n",
       "GSM904989_0026T_Exp.CEL      4.167279    6.086869      3.926209      4.000820   \n",
       "GSM904990_0042N_Exp.CEL      4.242397    6.336297      4.029943      4.483494   \n",
       "...                               ...         ...           ...           ...   \n",
       "GSM305112.CEL                3.623140    7.124676      3.767614      4.018204   \n",
       "GSM305113.CEL                3.725265    6.931847      4.055042      4.138074   \n",
       "GSM305114.CEL                3.799916    6.136671      3.736061      3.608018   \n",
       "GSM305115.CEL                4.193258    6.850084      4.166192      4.287220   \n",
       "GSM305116.CEL                3.898903    6.643763      3.875280      3.616558   \n",
       "\n",
       "Symbol                   1552563_a_at  1552607_at  ...    ZWILCH     ZWINT  \\\n",
       "GSM904985_0014N_Exp.CEL      4.406512    4.929491  ...  6.492485  6.470051   \n",
       "GSM904986_0014T_Exp.CEL      3.826696    4.338781  ...  6.286828  7.491875   \n",
       "GSM904988_0026N_Exp.CEL      4.026041    4.973174  ...  5.748054  6.875448   \n",
       "GSM904989_0026T_Exp.CEL      3.821508    4.645940  ...  6.031481  7.578293   \n",
       "GSM904990_0042N_Exp.CEL      4.035602    4.798371  ...  5.964617  6.630055   \n",
       "...                               ...         ...  ...       ...       ...   \n",
       "GSM305112.CEL                4.466480    4.825818  ...  6.311605  8.180971   \n",
       "GSM305113.CEL                4.350811    4.730193  ...  6.414612  7.152420   \n",
       "GSM305114.CEL                3.942664    4.627653  ...  6.173863  8.952424   \n",
       "GSM305115.CEL                3.860280    4.833088  ...  6.805837  8.734431   \n",
       "GSM305116.CEL                3.925251    4.884328  ...  7.429347  7.343067   \n",
       "\n",
       "Symbol                       ZXDA      ZXDB      ZXDC    ZYG11A    ZYG11B  \\\n",
       "GSM904985_0014N_Exp.CEL  6.311163  5.141446  6.531090  5.263501  8.519608   \n",
       "GSM904986_0014T_Exp.CEL  6.998897  5.153848  6.105484  5.527082  8.349207   \n",
       "GSM904988_0026N_Exp.CEL  6.024658  5.197408  6.384318  5.878829  9.062821   \n",
       "GSM904989_0026T_Exp.CEL  6.753043  5.233582  6.783810  5.051750  8.167005   \n",
       "GSM904990_0042N_Exp.CEL  5.452919  4.918226  5.866362  6.279175  8.704949   \n",
       "...                           ...       ...       ...       ...       ...   \n",
       "GSM305112.CEL            6.332484  5.163339  6.301880  5.545485  7.943559   \n",
       "GSM305113.CEL            5.885697  5.171652  6.137954  5.507664  8.010172   \n",
       "GSM305114.CEL            6.582268  5.146336  6.318175  5.391035  8.517406   \n",
       "GSM305115.CEL            6.018990  5.350754  6.120509  5.969984  8.396805   \n",
       "GSM305116.CEL            6.964285  5.533286  6.539190  5.024315  8.618043   \n",
       "\n",
       "Symbol                        ZYX     ZZEF1      ZZZ3  \n",
       "GSM904985_0014N_Exp.CEL  7.424008  6.131245  7.506218  \n",
       "GSM904986_0014T_Exp.CEL  8.321408  6.346957  7.314657  \n",
       "GSM904988_0026N_Exp.CEL  7.490204  6.189373  7.569571  \n",
       "GSM904989_0026T_Exp.CEL  7.381080  6.078681  7.563371  \n",
       "GSM904990_0042N_Exp.CEL  7.301923  6.059809  7.520830  \n",
       "...                           ...       ...       ...  \n",
       "GSM305112.CEL            7.335683  6.315128  7.418185  \n",
       "GSM305113.CEL            7.625686  5.855927  6.761258  \n",
       "GSM305114.CEL            8.858774  5.960262  7.712653  \n",
       "GSM305115.CEL            8.194778  6.022666  7.640047  \n",
       "GSM305116.CEL            7.417723  5.797618  7.454214  \n",
       "\n",
       "[347 rows x 32927 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_symbol_data = pd.read_csv(\"C:/Users/TEJINDER SINGH/Downloads/New folder (13)/kidney_data_gene_names.csv\")\n",
    "gene_data = merge_common_genes(gene_symbol_data)\n",
    "gene_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature selection\n",
    "1. Differential expression analysis\n",
    "2. Recursive feature elimination\n",
    "3. Checking correlation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_DEG(p_value, dataset, log_FC):\n",
    "    \"\"\"\n",
    "    Finds differentially expressed genes with the given p_value and fold_change\n",
    "    \n",
    "    Args:\n",
    "    p_value - int value for specifying the p value for stastical significance\n",
    "    dataset - dataset on which the differential expression analysis is to be performed\n",
    "    fold_change - int value depicting fold change\n",
    "    \n",
    "    Returns:\n",
    "    A dataset containing genes that are differentially expressed with given p value and fold change\n",
    "    \"\"\"\n",
    "    \n",
    "    data[\"sample_type\"] = labels\n",
    "    groups = data.groupby(\"sample_type\")\n",
    "    normal_sample = groups.get_group(\"normal\")\n",
    "    tumor_sample = groups.get_group(\"tumor\")\n",
    "    tumor_sample = tumor_sample.drop([\"sample_type\"], axis = 1)\n",
    "    normal_sample = normal_sample.drop([\"sample_type\"], axis = 1)\n",
    "\n",
    "    columns = normal_sample.columns\n",
    "    u_data = []\n",
    "    t_data = []\n",
    "  \n",
    "    for i in range(0,(len(columns)-1)):\n",
    "        a = normal_sample[columns[i]]\n",
    "        b = tumor_sample[columns[i]]\n",
    "        # T-test    \n",
    "        t,p = ttest_ind(a,b)\n",
    "        if p<p_value:\n",
    "            t_data.append(columns[i])\n",
    "\n",
    "    common_genes = list(set(u_data).intersection(t_data))\n",
    "\n",
    "    p_data = pd.DataFrame(kidney_data, columns = common)\n",
    "    p_data[\"samples\"] = labels\n",
    "\n",
    "    group = p_data.groupby(\"samples\")\n",
    "    n = group.get_group(\"normal\")\n",
    "    t = group.get_group(\"tumor\")\n",
    "    t.drop(\"samples\", axis = 1, inplace = True)\n",
    "    n.drop(\"samples\", axis = 1, inplace = True)\n",
    "\n",
    "    fold = []\n",
    "    for i in t.columns:\n",
    "        f = t[i].mean() - n[i].mean()\n",
    "        if f < -(log_FC) or f >(log_FC):\n",
    "            fold.append(i)\n",
    "    fold2 = pd.DataFrame(kidney_data, columns = fold)\n",
    "  \n",
    "    return fold2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Recursive feature elimination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def svc_rfe_cv(dataset, label):\n",
    "    \"\"\"\n",
    "    Performing recursive feature elimination using support vector classifier with 10 fold cross validation\n",
    "    \n",
    "    Args:\n",
    "    dataset - training data\n",
    "    label - trainig data labels\n",
    "    \n",
    "    Returns:\n",
    "    A  list of most informative columns according to SVC_RFE\n",
    "    \"\"\"\n",
    "    estimator = SVC(kernel=\"linear\")\n",
    "    selector = RFECV(estimator, min_features_to_select=100, step=1, cv = 10)\n",
    "    selector = selector.fit(dataset, label)\n",
    "    output = dataset[dataset.columns[selector.get_support()]]\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Performing Differential expression analysis and recursive feature elimination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deg_data = get_DEG(0.0001,data, 1)\n",
    "training_data = svc_rfe_cv(deg_data, labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr = training_data.corr()\n",
    "fig, ax = plt.subplots(figsize=(30, 30))\n",
    "colormap = sns.diverging_palette(220, 10, as_cmap=True)\n",
    "sns.heatmap(corr,vmin=-1, vmax=1, cmap=colormap)\n",
    "plt.xticks(range(len(corr.columns)), corr.columns);\n",
    "plt.yticks(range(len(corr.columns)), corr.columns)\n",
    "plt.savefig('C:/Users/TEJINDER SINGH/Downloads/New folder (13)/heatmap.png')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Single experiment dataset\n",
    "Extracting the data of a single experiment from the meta-dataset. The extracted data include 144 samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indx = []\n",
    "for i in training_data.index:\n",
    "    if i.startswith(\"GSM130\"):\n",
    "        indx.append(i)\n",
    "        \n",
    "training_data['lbl'] = labels\n",
    "single_train_set = training_data.T[indx]\n",
    "single_train_set = single_train_set.T\n",
    "single_train_set        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "single_lbl = single_train_set.pop(\"lbl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_test_data(path):\n",
    "    \"\"\"\n",
    "    Reads and load the test dataset, extract selected features from the test dataset\n",
    "    and generate binary labels from test dataset labels\n",
    "    \n",
    "    Args:\n",
    "    path - path to the Processed test dataset\n",
    "    \n",
    "    Returns:\n",
    "    testing_data - Final test dataset\n",
    "    test_labels - Labels for the test dataset\n",
    "    test_labels_binary - labels with 0 (for normal sample) and 1 (for tumour samples)\n",
    "    \"\"\"\n",
    "    test_data = pd.read_csv(path)\n",
    "    testing_data = test_data[training_data.columns]\n",
    "    testing_data = testing_data.set_index(test_data.index)\n",
    "    test_labels = test_data.pop(\"labels\")\n",
    "\n",
    "    test_labels_binary =[]\n",
    "    for i in test_labels:\n",
    "        if i == \"tumor\":\n",
    "            test_labels_binary.append(1)\n",
    "        else:\n",
    "            test_labels_binary.append(0)\n",
    "    return testing_data, test_labels,test_labels_binary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Split data into training and testing data with 1:1 ratio "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train, X_test, y_train, y_test = train_test_split(training_data, labels, \n",
    "                                                    test_size = 0.5,  \n",
    "                                                    random_state = 42) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Loading Test dataset\n",
    "#### Test dataset was processed using bioconductor package for R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_set_path = \"C:/Users/TEJINDER SINGH/Downloads/New folder (13)/test_gse66272.csv\"\n",
    "testing_data, test_labels, test_labels_binary = get_test_data(test_set_path)\n",
    "\n",
    "testing_data[\"label\"] = test_labels\n",
    "testing_data[\"binary\"] = test_labels_binary\n",
    "X_test[\"label\"] = y_test\n",
    "X_test[\"binary\"] = ytest\n",
    "\n",
    "test_df = pd.concat([testing_data, X_test])\n",
    "test_binary_labels = test_df.pop(\"binary\")\n",
    "test_lab = test_df.pop(\"label\")\n",
    "test_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables\n",
    "\n",
    "### Training data \n",
    " training_data - Training meta-dataset (174 samples),                                                       \n",
    "   labels        - Labels for meta-dataset\n",
    "\n",
    " single_train_set - Training data from single experiment,                                                                 \n",
    "   single_lbl       - Labels for single experiment\n",
    "   \n",
    "### Test data\n",
    " test_df - Testing meta-dataset (228 samples),                                                                     \n",
    " test_lab - labels for test meta-dataset,                                                                              \n",
    " test_binary_labels - binary labels (\"0\" for normal sample and \"1\" for tumour sample)\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hyperparameter optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_best_model(classifier, parameters):\n",
    "    \"\"\"\n",
    "    Performing Grid Search with 5 fold cross validation to find the best hyperparameters\n",
    "    for machine learning models\n",
    "    \n",
    "    Args:\n",
    "    classifiers - Machine learning classifier\n",
    "    parameters - dictionary of a range of hyperparameters for a given classifier\n",
    "    \n",
    "    Returns \n",
    "    dictionary of best parameters,\n",
    "    Machine learning classifier with best score\n",
    "    \"\"\"\n",
    "    CV = GridSearchCV(classifier, param_grid=parameters)\n",
    "    CV.fit(training_data, labels)\n",
    "    return CV.best_params_, CV.best_estimator_\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Choosing classifiers and parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#XGB\n",
    "xgb = XGBClassifier(learning_rate=0.001, objective='binary:logistic',tree_method='gpu_hist')\n",
    "param_grid_xgb = {'min_child_weight': [1,3,5],\n",
    "                  'gamma': [0.5, 1, 2],\n",
    "                  'max_depth': [3,4,6],\n",
    "                  'n_estimators': [100,300,500,700]}\n",
    "\n",
    "#Decision Tree\n",
    "dtc= DecisionTreeClassifier()\n",
    "param_grid_dtc = {'max_features': ['auto', 'sqrt', 'log2'],\n",
    "                 'min_samples_split': [2,3,4,5,6,7,8,9,10,11,12,13,14,15], \n",
    "                 'min_samples_leaf':[1,2,3,4,5,6,7,8,9,10,11]}\n",
    "\n",
    "#Logistic regression\n",
    "lr = LogisticRegression(penalty='l2')\n",
    "param_grid_lr = {'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000],\n",
    "                 'tol':[0000.1,0.0001,000.1,00.1,0.1]}\n",
    "\n",
    "#Random Forest\n",
    "rfc = RandomForestClassifier()\n",
    "param_grid_rfc = {'n_estimators': [100, 200, 300, 400, 500],\n",
    "                 'max_features': ['auto', 'sqrt', 'log2'],\n",
    "                 'max_depth' : [4,5,6,7,8],\n",
    "                 'criterion' :['gini', 'entropy']}\n",
    "\n",
    "#Support Vector Machine\n",
    "svm = SVC(probability=True)\n",
    "param_grid_svm = {'C': [0.001, 0.01, 0.1, 1, 10],\n",
    "                  'gamma' : ['scale', 'auto']}\n",
    "\n",
    "\n",
    "#ADaboost\n",
    "adb = AdaBoostClassifier()\n",
    "param_grid_adb = {'base_estimator': [RandomForestClassifier(),DecisionTreeClassifier()],\n",
    "                  'n_estimators' : [10,20,30,40,50,60,70,80,90,100,150,200]}\n",
    "\n",
    "#GPC\n",
    "gpc = GaussianProcessClassifier(kernel=DotProduct(1.0))\n",
    "param_grid_gpc = {'kernel':[DotProduct(i) for i in [0.2, 0.5, 1,2,3,5]]  + [RBF(i) for i in [0.2, 0.5, 1,2,3,5]]}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "performing GridSearchCV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xgb_best_parameters, xgb_best_model = get_best_model(xgb, param_grid_xgb)\n",
    "dtc_best_parameters, dtc_best_model = get_best_model(dtc, param_grid_dtc)\n",
    "lr_best_parameters, lr_best_model   = get_best_model(lr, param_grid_lr)\n",
    "rfc_best_parameters, rfc_best_model = get_best_model(rfc, param_grid_rfc)\n",
    "svm_best_parameters, svm_best_model = get_best_model(svm, param_grid_svm)\n",
    "adb_best_parameters, adb_best_model = get_best_model(adb, param_grid_adb)\n",
    "gpc_best_parameters, gpc_best_model = get_best_model(gpc, param_grid_gpc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training\n",
    "## Adding the classifiers with the best parameters to a list "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ML models for meta-dataset\n",
    "meta_models = []\n",
    "meta_models.append(('LR', LogisticRegression(C=0.1, tol=0.1,)))\n",
    "meta_models.append(('LDA', LinearDiscriminantAnalysis()))\n",
    "meta_models.append(('QDA', QuadraticDiscriminantAnalysis()))\n",
    "meta_models.append(('DT', DecisionTreeClassifier(min_samples_leaf=4, min_samples_split=13,splitter='best')))\n",
    "meta_models.append(('NB', GaussianNB()))\n",
    "meta_models.append(('SVM', SVC(C=10, gamma='scale', kernel='rbf',probability=True)))\n",
    "meta_models.append(('ADB', AdaBoostClassifier(n_estimators=20)))\n",
    "meta_models.append(('RF', RandomForestClassifier(criterion='gini', max_depth=6, min_samples_leaf=1, min_samples_split=2,\n",
    "                       n_estimators=100)))\n",
    "meta_models.append(('GPC', GaussianProcessClassifier(kernel=RBF(length_scale=1))))\n",
    "meta_models.append(('XGB', XGBClassifier(booster='gbtree', colsample_bylevel=1,\n",
    "                    learning_rate=0.001, max_depth=6,\n",
    "                    min_child_weight=5,n_estimators=700,\n",
    "                    objective='binary:logistic')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ML models for single experiment dataset\n",
    "single_models = []\n",
    "single_models.append(('LR', LogisticRegression(C=0.1, tol=0.1,)))\n",
    "single_models.append(('LDA', LinearDiscriminantAnalysis()))\n",
    "single_models.append(('QDA', QuadraticDiscriminantAnalysis()))\n",
    "single_models.append(('DT', DecisionTreeClassifier(min_samples_leaf=4, min_samples_split=13,splitter='best')))\n",
    "single_models.append(('NB', GaussianNB()))\n",
    "single_models.append(('SVM', SVC(C=10, gamma='scale', kernel='rbf',probability=True)))\n",
    "single_models.append(('ADB', AdaBoostClassifier(n_estimators=20)))\n",
    "single_models.append(('RF', RandomForestClassifier(criterion='gini', max_depth=6, min_samples_leaf=1, min_samples_split=2,\n",
    "                       n_estimators=100)))\n",
    "single_models.append(('GPC', GaussianProcessClassifier(kernel=RBF(length_scale=1))))\n",
    "single_models.append(('XGB', XGBClassifier(booster='gbtree', colsample_bylevel=1,\n",
    "                     learning_rate=0.001, max_depth=6,\n",
    "                     min_child_weight=5,n_estimators=700,objective='binary:logistic')))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Converting binary labels into 0 and 1\n",
    "ytest =[]\n",
    "for i in y_test:\n",
    "    if i == \"tumor\":\n",
    "        ytest.append(1)\n",
    "    else:\n",
    "        ytest.append(0)\n",
    "\n",
    "# meta Model_training\n",
    "for name, model in meta_models:\n",
    "    model.fit(X_train, y_train)\n",
    "\n",
    "# single Model_training\n",
    "for name, model in single_models:\n",
    "    model.fit(single_train_set, single_lbl)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# After Data processing and Training machine learning classifiers the models were tested on the test dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_results(name,model, test_data, test_label, binary_labels, path):\n",
    "    \"\"\"\n",
    "    This function makes prediction on the test dataset and returns the results\n",
    "    \n",
    "    Args:\n",
    "    name - String ,name of the classifier\n",
    "    model - classifier object\n",
    "    test_dataset - test dataframe\n",
    "    test_label - labels for the test dataframe\n",
    "    binary_labels - labels for the test data set in binary form (\"0\" for normal sample and \"1\" for tumour sample)\n",
    "    path - path to save confusion matrix\n",
    "    \n",
    "    Returns:\n",
    "    clf_r - classificationreport\n",
    "    accuracy - accuracy of the classifier on test dataset\n",
    "    fpr1 - False positive rate\n",
    "    tpr1 - True positive rate\n",
    "    roc_auc - Area under the ROC curve\n",
    "    precision - Precision of classifier on test dataset\n",
    "    recall - Recall of the classifier on test dataset\n",
    "    f1 - F1 score\n",
    "    log_los - Log loss\n",
    "    ap - Average precision\n",
    "    \"\"\"\n",
    "    \n",
    "    prediction = model.predict(test_data)\n",
    "    clf_r = classification_report(test_label, prediction)\n",
    "    accuracy = accuracy_score(test_label, prediction)\n",
    "    #print(clf_r,\"\\n\",f\"Accuracy = {accuracy*100}%\")\n",
    "    #plot_roc(model, testing_data, binary_labels)\n",
    "    \n",
    "    pred_binary = []\n",
    "    for i in prediction:\n",
    "        if i == \"tumor\":\n",
    "            pred_binary.append(1)\n",
    "        else:\n",
    "            pred_binary.append(0)\n",
    "    probs = model.predict_proba(test_data)[:,1]\n",
    "    #print(probs)        \n",
    "    fpr1, tpr1, _ = roc_curve(binary_labels, probs)\n",
    "    roc_auc = auc(fpr1, tpr1)\n",
    "    log_los = log_loss(binary_labels, probs)\n",
    "    \n",
    "    precision, recall, _ = precision_recall_curve(binary_labels, probs)\n",
    "    f1 = f1_score(binary_labels, pred_binary) \n",
    "    label = [\"tumor\", \"normal\"]\n",
    "    \n",
    "    cm = confusion_matrix(test_label, prediction,labels = label)\n",
    "    print(cm)\n",
    "    cm_display = ConfusionMatrixDisplay(cm,display_labels = label).plot(cmap = \"Reds\")\n",
    "    plt.savefig(f'{path}{name}.png')\n",
    "    plt.show()\n",
    "    ap = average_precision_score(binary_labels, pred_binary)\n",
    "    #wrong predictions indices\n",
    "    #indices = [i for i in range(len(test_label)) if test_label[i] != prediction[i]]\n",
    "    \n",
    "    #try:\n",
    "    #    print(\"Getting features\")\n",
    "    #    model1 = SelectFromModel(model, prefit=True)\n",
    "    #    feature_idx = model1.get_support()\n",
    "                #print(feature_idx)\n",
    "    #except:\n",
    "    #    feature_idx = 0\n",
    "    #    print(\"No features available\")\n",
    "    #data.rename(columns={\"Symbol\":i}, inplace=True)\n",
    "    #feat = pd.concat([feat,data],axis = 1)\n",
    "\n",
    "    return clf_r, accuracy, fpr1, tpr1, roc_auc, precision, recall,  f1, log_los, ap\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pr_curve(result, path):\n",
    "    \"\"\" \n",
    "    Plots the Precision recall curve\n",
    "    \n",
    "    Args:\n",
    "    result - Result dataframe containing the following columns\n",
    "             'recall'\n",
    "             'precision'\n",
    "             'AP'\n",
    "    path - path to save the PR plot\n",
    "    \n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "\n",
    "    fig = plt.figure(figsize=(8,6))\n",
    "\n",
    "    for i in result.index:\n",
    "        plt.plot(result.loc[i]['recall'], \n",
    "                 result.loc[i]['precision'], \n",
    "                 label = f\"{i}, AP={result.loc[i]['AP']}\")\n",
    "\n",
    "    #plt.plot([0,1], [0,1], color='orange', linestyle='--')\n",
    "    plt.xticks(np.arange(0.0, 1.1, step=0.1))\n",
    "    plt.xlabel(\"recall\", fontsize=15)\n",
    "\n",
    "    plt.yticks(np.arange(0.0, 1.15, step=0.1))\n",
    "    plt.ylabel(\"precision\", fontsize=15)\n",
    "\n",
    "    plt.title('Precision Recall Curve', fontweight='bold', fontsize=15)\n",
    "    plt.legend(prop={'size':13}, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    ax = fig.add_subplot(111)\n",
    "    lgd = ax.legend(prop={'size':13}, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    plt.savefig(path,bbox_extra_artists=(lgd,), bbox_inches='tight')\n",
    "    plt.show()\n",
    "    \n",
    "def plot_roc_curve(result, path):\n",
    "    \"\"\" \n",
    "    Plots the ROC curve\n",
    "    \n",
    "    Args:\n",
    "    result - Result dataframe containing the following columns\n",
    "             'fpr'\n",
    "             'tpr'\n",
    "             'AUC'\n",
    "    path - path to save the ROC plot\n",
    "    \n",
    "    Returns:\n",
    "    None\n",
    "    \"\"\"\n",
    "    \n",
    "    fig = plt.figure(figsize=(8,6))\n",
    "\n",
    "    for i in result.index:\n",
    "        plt.plot(result.loc[i]['fpr'], \n",
    "                 result.loc[i]['tpr'], \n",
    "                 label = f\"{i}, AUC={result.loc[i]['auc']}\" )\n",
    "\n",
    "    #plt.plot([0,1], [0,1], color='orange', linestyle='--')\n",
    "    plt.xticks(np.arange(0.0, 1.1, step=0.1))\n",
    "    plt.xlabel(\"Flase Positive Rate\", fontsize=15)\n",
    "\n",
    "    plt.yticks(np.arange(0.0, 1.1, step=0.1))\n",
    "    plt.ylabel(\"True Positive Rate\", fontsize=15)\n",
    "\n",
    "    plt.title('ROC Curve', fontweight='bold', fontsize=15)\n",
    "    plt.legend(prop={'size':13}, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    ax = fig.add_subplot(111)\n",
    "    lgd = ax.legend(prop={'size':13}, loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "    \n",
    "    plt.savefig(path,bbox_extra_artists=(lgd,), bbox_inches='tight')\n",
    "    plt.show()    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Making predictions and result evaluations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-1-322ad152da15>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[0mclf_report_test\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 2\u001b[1;33m \u001b[0mresult_table_test\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'classifiers'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'accuracy'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'fpr'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'tpr'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'auc'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'precision'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'recall'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'f1 score'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'AP'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      3\u001b[0m \u001b[0mcm_path\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"C:/Users/TEJINDER SINGH/Downloads/Project results/\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmodel\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mmodels\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# Meta dataset model evaluation\n",
    "meta_clf_report_test = []\n",
    "meta_result_table_test = pd.DataFrame(columns=['classifiers','accuracy', 'fpr','tpr','auc', 'precision', 'recall', 'f1 score', 'AP'])\n",
    "meta_cm_path = \"C:/Users/TEJINDER SINGH/Downloads/Project results/meta_\"\n",
    "\n",
    "for name, model in meta_models:\n",
    "    print(name)\n",
    "    clf_r, acc, fpr, tpr, roc_auc, precision, recall,  f1 ,log_los,ap = get_results(name,model,test_df,lab, binary_labels, meta_cm_path)\n",
    "    meta_clf_report_test.append((name,clf_r))\n",
    "    meta_result_table_test  = meta_result_table_test.append({'classifiers': name,\n",
    "                                                       'accuracy' : f\"{acc*100}%\",\n",
    "                                                       'fpr':fpr, \n",
    "                                                       'tpr':tpr, \n",
    "                                                       'auc':roc_auc,\n",
    "                                                       'precision' : precision,\n",
    "                                                       'recall' : recall,\n",
    "                                                       'f1 score' : f1,\n",
    "                                                      'log_loss' : log_los,\n",
    "                                                      'AP' : ap}, ignore_index=True)\n",
    "\n",
    "meta_result_table_test.set_index('classifiers', inplace=True)\n",
    "meta_result_table_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#single_models evaluation\n",
    "single_clf_report_test = []\n",
    "single_result_table_test = pd.DataFrame(columns=['classifiers','accuracy', 'fpr','tpr','auc', 'precision', 'recall', 'f1 score', 'AP'])\n",
    "single_cm_path = \"C:/Users/TEJINDER SINGH/Downloads/Project results/single_\"\n",
    "\n",
    "for name, model in single_models:\n",
    "    print(name)\n",
    "    clf_r, acc, fpr, tpr, roc_auc, precision, recall,  f1 ,log_los,ap = get_results(name,model,test_df,lab, binary_labels, single_cm_path)\n",
    "    single_clf_report_test.append((name,clf_r))\n",
    "    single_result_table_test  = result_table_test.append({'classifiers': name,\n",
    "                                                       'accuracy' : f\"{acc*100}%\",\n",
    "                                                       'fpr':fpr, \n",
    "                                                       'tpr':tpr, \n",
    "                                                       'auc':roc_auc,\n",
    "                                                       'precision' : precision,\n",
    "                                                       'recall' : recall,\n",
    "                                                       'f1 score' : f1,\n",
    "                                                      'log_loss' : log_los,\n",
    "                                                      'AP' : ap}, ignore_index=True)\n",
    "\n",
    "single_result_table_test.set_index('classifiers', inplace=True)\n",
    "single_result_table_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot Precision recall curve meta_dataset\n",
    "path_pr = 'C:/Users/TEJINDER SINGH/Downloads/Project results/meta_PR_curve.png'\n",
    "plot_pr_curve(meta_result_table_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot Precision recall curve single_dataset\n",
    "path_pr = 'C:/Users/TEJINDER SINGH/Downloads/Project results/single_PR_curve.png'\n",
    "plot_pr_curve(single_result_table_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ROC curve meta_dataset\n",
    "path_roc = 'C:/Users/TEJINDER SINGH/Downloads/Project results/meta_ROC_curve.png'\n",
    "plot_roc_curve(meta_result_table_test, path_roc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ROC curve single_dataset\n",
    "path_roc = 'C:/Users/TEJINDER SINGH/Downloads/Project results/single_ROC_curve.png'\n",
    "plot_roc_curve(single_result_table_test, path_roc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the result table meta_data\n",
    "result_output = meta_result_table_test[['accuracy', 'auc', 'f1 score', 'AP', 'log_loss']]\n",
    "result_output.to_csv('C:/Users/TEJINDER SINGH/Downloads/Project results/result_single_exp_data.csv',index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the result table single data\n",
    "result_output = single_result_table_test[['accuracy', 'auc', 'f1 score', 'AP', 'log_loss']]\n",
    "result_output.to_csv('C:/Users/TEJINDER SINGH/Downloads/Project results/result_single_exp_data.csv',index = True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
